logbook-api/src/main/java/org/zalando/logbook/DefaultHttpHeaders.java

```diff
@@ -43,34 +44,38 @@ public HttpHeaders update(
             final String name,
             final Collection<String> values) {
 
-        return withHeaders(
-                headers.assoc(name, immutableCopy(values)));
+        TreeMap<String, List<String>> updatedHeaders = associate(headers, name, immutableCopy(values));
+        return headers.equals(updatedHeaders) ? this : withHeaders(updatedHeaders);
     }
 
     @Override
     public HttpHeaders delete(final Collection<String> names) {
-        return withHeaders(
-                Fold.fold(names, headers, ImSortedMap::without));
+        TreeMap<String, List<String>> updatedHeaders = delete(headers, names);
+        return headers.equals(updatedHeaders) ? this : withHeaders(updatedHeaders);
     }
 
     public static <T> List<T> immutableCopy(final Collection<T> values) {
-        if (values instanceof ImListWithToString) {
-            return (List<T>) values;
-        }
-        return new ImListWithToString<>(ofIter(values));
+        return Collections.unmodifiableList(new ArrayList<>(values));
     }
 
-    @AllArgsConstructor
-    private static final class ImListWithToString<T> extends AbstractList<T> {
-
-        private final ImList<T> list;
-
-        @Delegate
-        @SuppressWarnings("unused")
-        private List<T> delegate() {
-            return list;
-        }
+    private static <K, V> TreeMap<K, V> associate(final TreeMap<K, V> original, final K key, final V value) {
+        Comparator<? super K> cmp = original.comparator();
+        Stream<Map.Entry<K, V>> tmpStream = Stream.of(new SimpleEntry<>(key, value));
+
+        return Stream.concat(original.entrySet().stream(), tmpStream)
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        entry -> cmp.compare(key, entry.getKey()) == 0 ? value : entry.getValue(),
+                        (v1, v2) -> v2,
+                        () -> new TreeMap<>(cmp)
+                ));
```

Comment:
Would this be the same as this? 

```suggestion
        TreeMap<K, V> tmpMap = new TreeMap<>(original);
        tmpMap.put(key, value);
        return tmpMap;
```

Reply:
I believe the first version is a complicated way of saying `tmpMap.putIfAbsent(key, value)` instead of just `put`.

Reply:
Ah, nevermind. I misread the old code. I think you're right.

Reply:
@kasmarian Yes, you are right. I'll simplify it and push again once I have time today :)

Reply:
@kasmarian The fix is applied now and I also simplified a bit here and there to make it more efficient. Thanks a lot for your suggestion. Would you mind reviewing again? 🙏